filmreel/src/reel.rs

```
@@ -0,0 +1,156 @@
+/// │ ┌───────── Frame type             // se
+/// │ │  ┌────── Reel name              // usr
+/// │ │  │   ┌─  Method name            // createuser
+/// ▼ ▼  ▼   ▼
+/// 01se.usr.createuser.fr.json
+///                     ▲
+///                     └─ Frame suffix // .fr.json
+/// ```
+///
+#[derive(PartialEq, Debug)]
+pub struct MetaFrame {
+    path: PathBuf,
+    name: String,
+    reel_name: String,
+    step: f32,
+    frame_type: FrameType,
+}
+
+impl TryFrom<PathBuf> for MetaFrame {
+    type Error = Box<dyn Error>;
+
+    fn try_from(p: PathBuf) -> Result<Self, Self::Error> {
+        let mut reel_parts: Vec<&str> = p
+            .file_name()
+            .map(|s| s.to_str())
+            .expect("file_name.to_str() error")
+            .map(|s| s.trim_end_matches(".fr.json"))
+            .map(|s| s.split('.').collect())
+            .expect("str split error");
+
+        let (seq, fr_type) = parse_sequence(reel_parts.remove(0))?;
+        let reel_name = String::from(reel_parts.remove(0));
+        let name = reel_parts.remove(0);
+
+        // only three indices should be present when split on '.'
+        assert!(reel_parts.is_empty());
+
+        Ok(Self {
+            path: p.clone(),
+            name: name.to_string(),
+            reel_name,
+            step: seq,
+            frame_type: fr_type,
+        })
+    }
+}
+
+fn parse_sequence(seq: &str) -> Result<(f32, FrameType), Box<dyn Error>> {
+    let mut seq_chars: Vec<char> = Vec::new();
+    let mut type_str: String = String::new();
+    for ch in seq.chars() {
+        match ch {
+            // [0-9]
+            ch if ch.is_ascii_digit() => {
+                seq_chars.push(ch);
+            }
+            // [A-Za-z]
+            ch if ch.is_ascii_alphabetic() => {
+                type_str.push(ch);
+            }
+            '_' => {
+                seq_chars.push('.');
+            }
+            _ => {
+                unimplemented!("{} is an invalid sequence char!", ch);
```

Q: since we're already returning an error from this function on failure, what's the rationale for doing an `unimplemented!()` here rather than returning an Error?

---

My (poor) rationale was that I wasn't sure what to do with no ascii chars and was just fence-sitting. I will explicitly return an err for any non ascii chars, that's for asking!